src/tests/core/blocks/test_ser_block.c

```diff
@@ -0,0 +1,192 @@
+  unsigned char *sered_header, *sered_header_2;
+
+  header = malloc(sizeof(BlockHeader));
+  _set_header(header);
+
+  sered_header = ser_blockheader_alloc(header);
+  desered_header = deser_blockheader_alloc(sered_header);
+
+  mu_assert(
+    "Timestamp isn't consistent after serialization",
+    header->timestamp == desered_header->timestamp
+  );
+  mu_assert(
+    "Nonce isn't consistent after serialization",
+    header->nonce == desered_header->nonce
+  );
+  mu_assert(
+    "Prev hash isn't consistent after serialization",
+    memcmp(
+      header->prev_header_hash,
+      desered_header->prev_header_hash,
+      BLOCK_HASH_LEN
+    ) == 0
+  );
+  mu_assert(
+    "All tx hash isn't consistent after serialization",
+    memcmp(
+      header->all_tx,
+      desered_header->all_tx,
+      ALL_TX_HASH_LEN
+    ) == 0
+  );
+
+  // Ensure that we don't have padding bytes
+  for (int i = 0; i < 10; i++) {
+    sered_header_2 = ser_blockheader_alloc(header);
+    mu_assert(
+      "Serialization of header isn't consistent",
+      memcmp(sered_header, sered_header_2, size_ser_blockheader())  == 0
+    );
+    free(sered_header_2);
+  }
+
+  free(header);
+  free(sered_header);
+  free(desered_header);
+
+  return NULL;
+}
+
+static char *test_ser_block(){
+  Block *block, *desered_block;
+  unsigned char *sered_block, *sered_block_2;
+
+  _fill_mempool();
+  block = create_block_alloc();
+  sered_block = ser_block_alloc(block);
+  desered_block = deser_block_alloc(sered_block);
+
+  mu_assert(
+    "Num Txs don't match",
+    block->num_txs == desered_block->num_txs
+  );
+  mu_assert(
+    "Header timestamp isn't consistent after serialization",
+    block->header.timestamp == desered_block->header.timestamp
+  );
+  mu_assert(
+    "Header nonce isn't consistent after serialization",
+    block->header.nonce == desered_block->header.nonce
+  );
+  mu_assert(
+    "Header prev hash isn't consistent after serialization",
+    memcmp(
+      block->header.prev_header_hash,
+      desered_block->header.prev_header_hash,
+      BLOCK_HASH_LEN
+    ) == 0
+  );
+  mu_assert(
+    "Header all tx hash isn't consistent after serialization",
+    memcmp(
+      block->header.all_tx,
+      desered_block->header.all_tx,
+      ALL_TX_HASH_LEN
+    ) == 0
+  );
+
+  unsigned char base_tx_hash[TX_HASH_LEN];
+  unsigned char deser_tx_hash[TX_HASH_LEN];
+  for(unsigned int i = 0; i < block->num_txs; i++){
+    hash_tx(base_tx_hash, block->txs[i]);
+    hash_tx(deser_tx_hash, desered_block->txs[i]);
+    mu_assert(
+      "Transaction hashes don't match -> Transaction serialization error",
+      memcmp(base_tx_hash, deser_tx_hash, TX_HASH_LEN) == 0
+    );
+  }
+
+  // Ensure that we don't have padding bytes
+  for (int i = 0; i < 10; i++) {
+    sered_block_2 = ser_block_alloc(block);
+    mu_assert(
+      "Serialization of block isn't consistent",
+      memcmp(sered_block, sered_block_2, size_ser_block(block)) == 0
```

Comment:
Won't there be different pointers in this that break the memcmp?

Reply:
This is comparing the serialized versions which are the raw data w/ no pointers.